src/LineSegment.cpp

```diff
@@ -1,339 +1,135 @@
-        double t = numer/denom;
-        if (! (t < 0 || t > 1)) {
-            return start - A*t;
+        // Check if both LineSegments are on the same infinite line.
+        Line checkLine = Line(*this);
+        if (!checkLine.isOnLine(line.start)) {
+            return std::nullopt;
         }
-    }
-    return std::nullopt;
-
-}
-
-// only returns a vector if there is a point intersection. If a segment intersects does not return anything
-std::optional<Vector2> LineSegment::intersects(const LineSegment &line) const {
-    Vector2 A = start - end;
-    Vector2 B = line.start - line.end;
-    Vector2 C = start - line.start;
-    double numer = C.cross(B);
-    double denom = A.cross(B);
-    if (denom != 0) {
-        double t = numer/denom;
-        double u = - A.cross(C)/denom;
-        if (! (t < 0 || t > 1) && ! (u < 0 || u > 1)) {
-            return start - A*t;
-        }
-    }
-    else if (numer == 0) {
-        double t0 = C.dot(A)/A.length2();
-        double t1 = t0 + B.dot(A)/A.length2();
-        //check the ends for point intersections
-        if ((t0 == 0 && t1 < 0) || (t1 == 0 && t0 < 0)) {
+        if (line.isOnFiniteLine(start)) {
             return start;
-        }
-        else if ((t0 == 1 && t1 > 1) || (t1 == 1 && t0 > 1)) {
+        } else if (line.isOnFiniteLine(end)) {
             return end;
+        } else if (isOnFiniteLine(line.start)) {
+            return line.start;
+        } else {
+            return std::nullopt;
         }
     }
-    return std::nullopt;
 }
 
-double LineSegment::distanceToLine(const Vector2 &point) const {
-    return (this->project(point) - point).length();
-}
+double LineSegment::distanceToLine(const Vector2 &point) const { return (project(point) - point).length(); }
 
-//same principle but now we do not necessarily have an orthogonal vector but just pick the closest point on the segment
 Vector2 LineSegment::project(const Vector2 &point) const {
-    Vector2 AB = direction();
-    Vector2 AP = point - start;
-    double t = AP.dot(AB)/length2();
-    if (t < 0) {
-        return Vector2(start);
+    if (isPoint()) {
+        return start;
     }
-    else if (t > 1) {
-        return Vector2(end);
+    Vector2 projection = Line(*this).project(point);
+    double t = Line::relativePosition(start, end, projection);
+    if (t < 0) {
+        return start;
+    } else if (t > 1) {
+        return end;
     }
-    return Vector2(start + AB*t);
+    return projection;
 }
 
 bool LineSegment::isOnLine(const Vector2 &point) const {
     if (isPoint()) {
         return (start == point || end == point);
     }
-    Vector2 A = end - start;
-    Vector2 B = point - start;
-    double cross = A.cross(B);
-    if (cross != 0) {
-        return false;
-    }
-    //check if the point is in between the two points
-    double dot = A.dot(B);
-    if (dot < 0) {
-        return false;
-    }
-    if (dot > A.dot(A)) {
-        return false;
+    if (Line(*this).isOnLine(point)) {
```

Comment:
So you are contructing this line inside of this line? or what

Reply:
True, I am re-using the functionality inside the Line class to prevent code clones